internal type corruption due to use of variadic args

The following program `sample.go` triggers a panic:

```go
package main
  
func variadic(s ...string) {}
    
func f(s string) { println(s + "truc") }
    
func main() { f("machin") }
```

Expected result:
```console
$ go run ./sample.go
machintruc
```

Got:
```console
$ yaegi ./sample.go
sample.go:7:15: panic
panic: reflect: call of reflect.Append on string Value [recovered]
        panic: reflect: call of reflect.Append on string Value

goroutine 1 [running]:
github.com/containous/yaegi/interp.runCfg.func1(0xc000174230, 0xc00021c200)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:107 +0x1e0
panic(0xc45500, 0xc00015c720)
        /usr/lib/go/src/runtime/panic.go:679 +0x1b2
reflect.flag.mustBe(...)
        /usr/lib/go/src/reflect/value.go:208
reflect.Append(0xc08c60, 0xc00015a680, 0x198, 0xc00016dcc0, 0x1, 0x1, 0x5e4365, 0xc00021c200, 0xc00016dc10)
        /usr/lib/go/src/reflect/value.go:2025 +0x200
github.com/containous/yaegi/interp.call.func4(0xc000174230, 0xc000218000)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:708 +0x994
github.com/containous/yaegi/interp.runCfg(0xc00021c200, 0xc000174230)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:112 +0x7a
github.com/containous/yaegi/interp.(*Interpreter).run(0xc000208000, 0xc000219b00, 0xc000174050)
        /home/marc/go/src/github.com/containous/yaegi/interp/run.go:93 +0x12c
github.com/containous/yaegi/interp.(*Interpreter).Eval(0xc000208000, 0xc000212000, 0x73, 0x273, 0xc000212000, 0x73, 0x0, 0xc0001558c0)
        /home/marc/go/src/github.com/containous/yaegi/interp/interp.go:299 +0x3f0
main.main()
        /home/marc/go/src/github.com/containous/yaegi/cmd/yaegi/yaegi.go:77 +0x2f9
```

The program works well just by commenting out the variadic func definition.

The problem is due to the way variadic attribute is used in the interpreted, and applies to all type instances rather than just the concerned object.

--- a/_test/variadic4.go
+++ b/_test/variadic4.go
@@ -1,0 +1,10 @@
+package main
+
+func variadic(s ...string) {}
+
+func f(s string) { println(s + "bar") }
+
+func main() { f("foo") }
+
+// Output:
+// foobar

--- a/interp/cfg.go
+++ b/interp/cfg.go
@@ -89,7 +89,7 @@ func (interp *Interpreter) cfg(root *node) ([]*node, error) {
 					case stringT:
 						ktyp = sc.getType("int")
 						vtyp = sc.getType("byte")
-					case arrayT:
+					case arrayT, variadicT:
 						ktyp = sc.getType("int")
 						vtyp = o.typ.val
 					}
@@ -259,9 +259,6 @@ func (interp *Interpreter) cfg(root *node) ([]*node, error) {
 				var typ *itype
 				if typ, err = nodeType(interp, sc, c.lastChild()); err != nil {
 					return false
-				}
-				if typ.variadic {
-					typ = &itype{cat: arrayT, val: typ}
 				}
 				for _, cc := range c.child[:len(c.child)-1] {
 					sc.sym[cc.ident] = &symbol{index: sc.add(typ), kind: varSym, typ: typ}
@@ -1584,7 +1581,7 @@ func variadicPos(n *node) int {
 		return -1
 	}
 	last := len(n.child[0].typ.arg) - 1
-	if n.child[0].typ.arg[last].variadic {
+	if n.child[0].typ.arg[last].cat == variadicT {
 		return last
 	}
 	return -1

--- a/interp/run.go
+++ b/interp/run.go
@@ -614,7 +614,7 @@ func call(n *node) {
 			if c.kind == basicLit {
 				var argType reflect.Type
 				if variadic >= 0 && i >= variadic {
-					argType = n.child[0].typ.arg[variadic].TypeOf()
+					argType = n.child[0].typ.arg[variadic].val.TypeOf()
 				} else {
 					argType = n.child[0].typ.arg[i].TypeOf()
 				}

--- a/interp/type.go
+++ b/interp/type.go
@@ -44,6 +44,7 @@ const (
 	uint64T
 	uintptrT
 	valueT
+	variadicT
 	maxT
 )
 
@@ -82,6 +83,7 @@ var cats = [...]string{
 	uint64T:     "uint64T",
 	uintptrT:    "uintptrT",
 	valueT:      "valueT",
+	variadicT:   "variadicT",
 }
 
 func (c tcat) String() string {
@@ -102,17 +104,16 @@ type structField struct {
 // itype defines the internal representation of types in the interpreter
 type itype struct {
 	cat        tcat          // Type category
-	field      []structField // Array of struct fields if StrucT or InterfaceT
+	field      []structField // Array of struct fields if structT or interfaceT
 	key        *itype        // Type of key element if MapT or nil
-	val        *itype        // Type of value element if ChanT, MapT, PtrT, AliasT or ArrayT
-	arg        []*itype      // Argument types if FuncT or nil
-	ret        []*itype      // Return types if FuncT or nil
+	val        *itype        // Type of value element if chanT, mapT, ptrT, aliasT, arrayT or variadicT
+	arg        []*itype      // Argument types if funcT or nil
+	ret        []*itype      // Return types if funcT or nil
 	method     []*node       // Associated methods or nil
 	name       string        // name of type within its package for a defined type
 	pkgPath    string        // for a defined type, the package import path
 	size       int           // Size of array if ArrayT
 	rtype      reflect.Type  // Reflection type if ValueT, or nil
-	variadic   bool          // true if type is variadic
 	incomplete bool          // true if type must be parsed again (out of order declarations)
 	untyped    bool          // true for a literal value (string or number)
 	sizedef    bool          // true if array size is computed from type definition
@@ -280,10 +281,11 @@ func nodeType(interp *Interpreter, sc *scope, n *node) (*itype, error) {
 		t.incomplete = t.val.incomplete
 
 	case ellipsisExpr:
-		if t, err = nodeType(interp, sc, n.child[0]); err != nil {
+		t.cat = variadicT
+		if t.val, err = nodeType(interp, sc, n.child[0]); err != nil {
 			return nil, err
 		}
-		t.variadic = true
+		t.incomplete = t.val.incomplete
 
 	case funcLit:
 		t, err = nodeType(interp, sc, n.child[2])
@@ -720,7 +722,7 @@ func (t *itype) Type(name string) reflect.Type {
 	}
 
 	switch t.cat {
-	case arrayT:
+	case arrayT, variadicT:
 		if t.size > 0 {
 			t.rtype = reflect.ArrayOf(t.size, t.val.Type(name))
 		} else {
@@ -774,7 +776,7 @@ func (t *itype) frameType() (r reflect.Type) {
 		panic(err)
 	}
 	switch t.cat {
-	case arrayT:
+	case arrayT, variadicT:
 		if t.size > 0 {
 			r = reflect.ArrayOf(t.size, t.val.frameType())
 		} else {

